File: datastore/api/routers/strategies.py
"""
datastore/api/routers/strategies.py

Owner: Platform / Strategy registry (A95)
Consumers: the /backtest-report frontend, the deploy page, and anything that
needs to know what a strategy IS rather than how one run of it performed.

GET /api/v1/strategies      -- every declared strategy, from strategy_registry
GET /api/v1/strategies/{key}
GET /api/v1/filters         -- every declared filter, from filter_registry

This is the endpoint that makes AGENTS.md invariant 5 real: backtest, API and
frontend read the same rows. Before it, the frontend derived strategy identity
and labels client-side from variant-id string parsing, and the backtest
imported TEMPLATES/build_category_presets/STRATEGY_CATALOG directly -- three
copies of the same facts, free to drift.

Point-in-time is preserved: `version` and `as_of` select a historical
definition, because a run executed against version 3 must not be explained
using version 5's rules.
"""

from datetime import date
from pathlib import Path
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field

from config.settings import BACKTEST_DUCKDB_PATH
from strategies.registry import (
    get_filter,
    get_strategy,
    list_filters,
    list_strategies,
    resolve_filters,
)

router = APIRouter(prefix="/api/v1", tags=["Strategy registry"])


class StrategyOut(BaseModel):
    strategy_key: str
    version: int
    channel: str
    name: str
    display_label: Optional[str] = None
    description: Optional[str] = None
    category: Optional[str] = None
    definition: Dict[str, Any] = Field(default_factory=dict)
    # Entry is an ORDERED LIST of predicates ({feature, op, value|feature2}),
    # uniform across all four channels; exit is a single policy object. The
    # asymmetry is real, not an oversight: entry conditions compose and their
    # order is meaningful, an exit policy is one choice with parameters.
    entry_criterion: List[Dict[str, Any]] = Field(default_factory=list)
    exit_criterion: Dict[str, Any] = Field(default_factory=dict)
    filter_ids: List[str] = Field(default_factory=list)
    status: str
    valid_from: Optional[date] = None
    valid_to: Optional[date] = Field(
        default=None,
        description="Null for the current version. A non-null value means this row was superseded.",
    )
    source_ref: Optional[str] = None


class StrategyListResponse(BaseModel):
    strategies: List[StrategyOut]
    total: int
    channels: Dict[str, int] = Field(
        default_factory=dict, description="Count per channel, for the UI's filter chips."
    )


class FilterOut(BaseModel):
    filter_id: str
    version: int
    name: str
    description: Optional[str] = None
    filter_type: str
    params_schema: Dict[str, Any] = Field(default_factory=dict)
    default_params: Dict[str, Any] = Field(default_factory=dict)
    applies_to_channels: List[str] = Field(default_factory=list)
    implementation_ref: Optional[str] = None
    status: str


class FilterListResponse(BaseModel):
    filters: List[FilterOut]
    total: int


class ResolvedFilter(BaseModel):
    filter_id: str
    params: Dict[str, Any]
    implementation_ref: Optional[str] = None


def _db_path() -> Path:
    return BACKTEST_DUCKDB_PATH


@router.get("/strategies", response_model=StrategyListResponse)
def list_strategies_endpoint(
    channel: Optional[str] = Query(
        None, description="momentum | technical | fundamental | ml"
    ),
    status: Optional[str] = Query(
        "active",
        description=(
            "active | draft | retired. Pass an empty string for every status, "
            "which is what an audit view wants."
        ),
    ),
) -> StrategyListResponse:
    rows = list_strategies(
        channel=channel,
        status=status or None,
        db_path=_db_path(),
    )
    counts: Dict[str, int] = {}
    for r in rows:
        counts[r["channel"]] = counts.get(r["channel"], 0) + 1
    return StrategyListResponse(
        strategies=[StrategyOut(**r) for r in rows],
        total=len(rows),
        channels=counts,
    )


@router.get(
    "/strategies/{strategy_key:path}/filters", response_model=List[ResolvedFilter]
)
def resolve_strategy_filters(strategy_key: str) -> List[ResolvedFilter]:
    """A strategy's filters with their parameters already resolved -- declared
    defaults with the strategy's own overrides applied on top. This is what a
    runner should execute, and what the UI should display, so neither has to
    re-implement the override precedence."""
    row = get_strategy(strategy_key, db_path=_db_path())
    if row is None:
        raise HTTPException(status_code=404, detail=f"No strategy {strategy_key!r}")
    resolved = resolve_filters(
        row.get("filter_ids") or [],
        (row.get("definition") or {}).get("filter_overrides"),
        db_path=_db_path(),
    )
    return [ResolvedFilter(**r) for r in resolved]


# Registered BEFORE the greedy /strategies/{key:path} route below: `:path`
# matches slashes, so it would otherwise swallow "/filters" and this endpoint
# would be unreachable (it 404s with the suffix treated as part of the key).
# FastAPI matches in registration order, so order here is behaviour.
@router.get("/strategies/{strategy_key:path}", response_model=StrategyOut)
def get_strategy_endpoint(
    strategy_key: str,
    version: Optional[int] = Query(
        None, description="A specific version. Defaults to the current one."
    ),
    as_of: Optional[date] = Query(
        None,
        description=(
            "The definition in force on this date. Use this rather than the "
            "current row when explaining a historical run: a run executed "
            "against version 3 must not be described using version 5's rules."
        ),
    ),
) -> StrategyOut:
    row = get_strategy(
        strategy_key, version=version, as_of=as_of, db_path=_db_path()
    )
    if row is None:
        raise HTTPException(
            status_code=404,
            detail=(
                f"No strategy {strategy_key!r}"
                + (f" at version {version}" if version else "")
                + (f" as of {as_of}" if as_of else "")
            ),
        )
    return StrategyOut(**row)


@router.get("/filters", response_model=FilterListResponse)
def list_filters_endpoint(
    channel: Optional[str] = Query(
        None, description="Only filters that apply to this channel."
    ),
) -> FilterListResponse:
    rows = list_filters(channel=channel, db_path=_db_path())
    return FilterListResponse(
        filters=[FilterOut(**r) for r in rows], total=len(rows)
    )


@router.get("/filters/{filter_id}", response_model=FilterOut)
def get_filter_endpoint(filter_id: str) -> FilterOut:
    row = get_filter(filter_id, db_path=_db_path())
    if row is None:
        raise HTTPException(status_code=404, detail=f"No filter {filter_id!r}")
    return FilterOut(**row)

File: frontend/src/shared/api/strategies.ts
/**
 * shared/api/strategies.ts
 *
 * Typed client for the strategy/filter registries (A95).
 *
 * WHY THIS FILE EXISTS
 * --------------------
 * The registries answer "what IS this strategy" — its label, its entry
 * predicates, its exit policy, the filters it declares. Until now the
 * frontend answered that question itself, by parsing variant-id strings and
 * branching per channel to decide what a run's setup meant. That made three
 * copies of the same facts (backtest imports, API rows, client-side parsing)
 * which were free to drift, and a deployed strategy's definition could differ
 * from the one that was backtested with nothing to catch it.
 *
 * These rows are the single source. Run METRICS still come from the backtest
 * report endpoints (shared/api/backtest.ts) — that split is deliberate: a
 * registry row describes the strategy, a run describes one execution of it.
 *
 * Point-in-time is preserved by the API: pass `version` (or `asOf`) to read
 * the definition as it stood when a given run executed, because explaining a
 * version-3 run with version-5 rules is exactly the drift this replaces.
 */

import { apiGet } from './client'

/** One predicate in a strategy's ordered entry criterion. */
export interface EntryPredicate {
  feature?: string
  op?: string
  value?: unknown
  feature2?: string
  [k: string]: unknown
}

export interface RegistryStrategy {
  strategy_key: string
  version: number
  channel: string
  name: string
  display_label: string | null
  description: string | null
  category: string | null
  definition: Record<string, unknown>
  /**
   * ORDERED list — order is meaningful and must not be re-sorted for display.
   * Exit is a single policy object rather than a list; that asymmetry is real
   * (entry conditions compose, an exit policy is one choice with parameters).
   */
  entry_criterion: EntryPredicate[]
  exit_criterion: Record<string, unknown>
  filter_ids: string[]
  status: string
  valid_from: string | null
  valid_to: string | null
  source_ref?: string | null
}

export interface RegistryFilter {
  filter_id: string
  version: number
  name: string
  description: string | null
  filter_type: string | null
  params_schema: Record<string, unknown>
  default_params: Record<string, unknown>
  applies_to_channels?: string[]
}

interface StrategyListResponse {
  strategies: RegistryStrategy[]
}

interface FilterListResponse {
  filters: RegistryFilter[]
}

/**
 * Every declared strategy, optionally narrowed by channel/status.
 *
 * `limit` is passed through rather than defaulted here: momentum alone has
 * ~3,100 rows, so a caller that wants a picker must page or filter by channel
 * deliberately instead of silently receiving a truncated list it believes is
 * complete.
 */
export async function listStrategies(params?: {
  channel?: string
  status?: string
  limit?: number
  offset?: number
}): Promise<RegistryStrategy[]> {
  const resp = await apiGet<StrategyListResponse>('/api/v1/strategies', params)
  return resp.strategies ?? []
}

/**
 * One strategy's definition.
 *
 * `version` selects a historical revision; omit it for the current one. Pass
 * the version the RUN recorded when explaining a run — see the module note.
 */
export async function getStrategy(
  strategyKey: string,
  params?: { version?: number; asOf?: string },
): Promise<RegistryStrategy> {
  return apiGet<RegistryStrategy>(`/api/v1/strategies/${encodeURIComponent(strategyKey)}`, {
    version: params?.version,
    as_of: params?.asOf,
  })
}

/** Every declared filter concept, one row per implementation. */
export async function listFilters(): Promise<RegistryFilter[]> {
  const resp = await apiGet<FilterListResponse>('/api/v1/filters')
  return resp.filters ?? []
}
